src/policy/rbf.cpp

```diff
@@ -182,3 +182,43 @@ std::optional<std::string> PaysForRBF(CAmount original_fees,
     }
     return std::nullopt;
 }
+
+std::optional<std::string> CheckMinerScores(CAmount replacement_fees,
+                                            int64_t replacement_vsize,
+                                            const CTxMemPool::setEntries& ancestors,
+                                            const CTxMemPool::setEntries& direct_conflicts,
+                                            const CTxMemPool::setEntries& original_transactions)
+{
+        const CFeeRate replacement_individual_feerate(replacement_fees, replacement_vsize);
+        // Ancestor feerate is the total modified fees divided by the total size. To get the
+        // ancestor feerate, add up all the individual modified fees and sizes. Don't try to use the
+        // cached ancestor fees and sizes because entries may have overlapping ancestors.
+        for (CTxMemPool::txiter it : ancestors) {
+            replacement_fees += it->GetModifiedFee();
+            replacement_vsize += it->GetTxSize();
+        }
+        const CFeeRate replacement_ancestor_feerate(replacement_fees, replacement_vsize);
+        // A package/transaction's ancestor feerate is not equivalent to the miner score; it may
+        // overestimate. Some subset of the ancestors could be included by itself if it has other
+        // high-feerate descendants or are themselves higher feerate than this package/transaction.
+        // For now, as a conservative estimate, use the minimum between the transaction's individual
+        // feerate and ancestor feerate.
+        const CFeeRate replacement_miner_score = std::min(replacement_individual_feerate, replacement_ancestor_feerate);
```

Comment:
Can we relax the `replacement_ancestor_feerate` for v3 transactions? What would be ideal would be to simply ignore unconfirmed v3 ancestors from this calculation (some kind of `replacement_non_v3_ancestor_feerate`), which I believe is ok from an incentives' point of view given the other restrictions on v3 packages.

Otherwise we will have the following issue:

- a commitment transaction pays 0 fees and weighs 19,000 vbytes (it has many HTLCs)
- an attacker publishes a CPFP v3 child that weighs 1,000 vbytes and has an individual feerate of e.g. 100 sat/byte
- the package feerate ends up being 5 sat/byte, and the package doesn't confirm
- if a feerate of 20 sat/byte for the package was enough to get the package confirmed, the honest participant would target that and thus create a CPFP v3 child with a feerate of 400 sat/byte
- but that would be rejected under the current rules, and the honest participant is forced to make the _package feerate_ be 100 sat/byte, which is extremely expensive

I'm realizing now that this is actually just another way of formulating @instagibbs's proposal to evict v3 siblings: when the direct conflict is a v3 transaction with an unconfirmed v3 parent, it would be great to only compare their individual feerates.

Reply:
To clarify, is the second v3 child (400sat/vbyte) replacing the first v3 child (100sat/vbyte) only? Assuming the commitment tx is already in the mempool. In that case, it would just be single transaction replacement so this rule isn't used. Just the individual feerates are compared, and the child is accepted. Maybe I'm misunderstanding the scenario though?

Reply:
I think we cannot do only individual replacement here, if the attacker publishes their commit + their child, and the honest participants wants to publish their own commit + their child (because for some reason they never receive the attacker's package in their mempool so cannot attach a CPFP to that commit).

So this is actually slightly different from the v3 sibling eviction problem, which as you point out probably just works :tm: with individual replacement rules already.

Reply:
I see okay so the situation is:

original transactions = commitment_A + cpfp_A
replacement transactions (package) = commitment_B + cpfp_B
- commitment_A: 19,000vB, 0sat.
    - individual feerate = 0sat/vB
    - ancestor feerate = 0sat/vB
- cpfp_A: 1000vbytes, 100,000sat.
    - individual feerate = 100sat/vB
    - ancestor feerate = 5sat/vB
- commitment_B: 19,000vB, 0sat.
    - individual feerate = 0sat/vB
    - ancestor feerate = 0sat/vB
- cpfp_B: 1000vB, 400,000sat.
    - individual feerate = 400sat/vB
    - ancestor feerate = 20sat/vB
package feerate = 20sat/vB

`replacement_miner_score` = minimum(package feerate, ancestor feerate) = 20sat/vB
the rule is that `replacement_miner_score` must be higher than
- the direct conflict's (commitment_A) indivdual feerate = 0sat/vB
- the original transactions' (commitment_A, cpfp_A) ancestor feerates = 0sat/vB and 5sat/vB

So this should be fine AFAICT

Reply:
That's great! I'll write a test for this in eclair anyway, thanks for clearing that up :+1:

Reply:
This rule is still sometimes problematic *if* you are doing batch fee bumping.

Let's imagine the case where we have two separate channels, symmetrical commitment transactions(for clarity of argument).

If commitment_A is in the miner's mempool, and Bob wants to fee bump commitment_A as well as CPFP commitment_B(different channel!) into the miner's mempool, Bob will have to pay the replaced cpfp_A's feerate for the entire CPFP commitment_B + cpfp_B package feerate.

In other words, batch bumping is probably not a good idea unless you know package RBF will evict the sponsored transaction directly.